### ISSUE TITLE
omnisharp--do-server-start reports 'Wrong number of arguments: set-process-filter, 1'

### ISSUE THREAD

**Issue by ronnieholm** [OmniSharp/omnisharp-emacs#369]
When I open a Hello World C# file in Emacs and start the OmniSharp server from within Emacs, pointing it to the directory of the csproj file, OmniSharp emits the following message:

    omnisharp--do-server-start: Wrong number of arguments: set-process-filter, 1

However, OmniSharp appears to start up because it populates with OmniServer buffer with project information about Hello World. But querying the state of OmniSharp, Emacs reports

    omnisharp: OmniSharp server not running. Start it with ‘omnisharp-start-omnisharp-server’ first

I'm on Emacs 25.2.1 (x86_64-w64-mingw32) on Windows 7, using omnisharp-win-x86 against a .NET Core 2.0 project.

**Comment by razzmatazz** [comment]
@ronnieholm mhm, do you have omnisharp-emacs installed from melpa or melpa-stable?

**Comment by ronnieholm** [comment]
I'm using melpa. Is melpa-stable required?

**Comment by ronnieholm** [comment]
Sorry. The reason why it wasn't working is that I used

    (use-package omnisharp
      :ensure t)

instead of

    (package-install 'omnisharp)

**Comment by razzmatazz** [comment]
Nope, I was just think what could have changed since then..

Good to know!

**Comment by razzmatazz** [comment]
@ronnieholm btw, could you paste the related configuration of omnisharp-emacs in your init.el so I can update the documentation? as other users might trip on this and I am using spacemacs myself..

**Comment by ronnieholm** [comment]
This is what I use:
```
;; initialize MELPA
(require 'package)
(setq package-enable-at-startup nil)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
(package-initialize)

(unless (package-installed-p 'omnisharp)
  (package-refresh-contents)
  (package-install 'omnisharp)

;; https://github.com/OmniSharp/omnisharp-emacs/issues/315
;; Use win7-x86 or Emacs will crash on server start. This'll happen
;; until the next Emacs release. More recent Emacs release is from
;; 2017-04-24 and fix was committed to Emacs on June 9, 2017.
(add-hook 'csharp-mode-hook 'omnisharp-mode)
(eval-after-load
 'company
 '(add-to-list 'company-backends 'company-omnisharp))
x86/Omnisharp.exe")
(setq omnisharp-server-executable-path "C:/Users/ronnie/Desktop/omnisharp-win-x86/Omnisharp.exe")
(setq omnisharp-debug t)
```

**Comment by lwiechec** [comment]
I also have the same issue. I am using Emacs 25.3 with omnisharp-emacs installed from sources (I am using el-get). I guess it means that the implementation of `set-process-filter` and `set-process-sentinel` changed in the latest Emacs... 

PS: I have also error about `-doto` function/macro not found; where does it come from?...

**Comment by razzmatazz** [comment]
@lwiechec -doto comes from https://melpa.org/#/dash (dash.el) package. 

could it be that `use-package` does not install the dependencies?

@lwiechec @ronnieholm Btw, since I am using `spacemacs` and not plain emacs, would someone be able to submit a PR that updates `README.md` so it clarifies this installation issue better? I.e. that you should use `package-install` vs `use-package`?

IIRC when using regular emacs, I was installing omnisharp via `M-x list-packages` which I suppose does `package-install`, right? And it did install the dependencies right for me.

**Comment by lwiechec** [comment]
thanks for the clarification - I did some *more* effort and found out about `dash.el`. Thanks.

I have to admit that my setup is a bit haphazard; using `package-install` or `use-package` is what most people use.

**Comment by phoenixanimations** [comment]
@razzmatazz When I do a clean install using `package-install` I get the error as well. Omnisharp works just fine, and I don't get any errors when running the server. This is the error I get when it compiles Omnisharp from Melpa:

`omnisharp-server-management.el:64:1:Error: Wrong type argument: listp, value`

Going to the line, it does seem to support lwiechec's theory that `set-process-filter` and `set-process-sentinel` implementation might have changed. Both functions seem to have only one argument when they require two.

**Comment by razzmatazz** [comment]
@phoenixanimations the arity problem is most probably due to a -`-doto` macro from `dash.el` missing on your installation due to package misconfiguration(?) — I will check later if there is something wrong with omnisharp-emacs package or if we need better configuration documentation

**Comment by razzmatazz** [comment]
can confirm that there is a difference between
 - `M-x package-list-packages` / `S omnisharp` / `I` / `x` (which installs things properly for me)
 - (package-install 'omnisharp) – which fails for some reason, maybe some dependencies are not installed properly this way.. need to investigate

**Comment by razzmatazz** [comment]
`(package-install 'omnisharp)` seems to do OK, if i do it in `*scratch*` after bootup

**Comment by phoenixanimations** [comment]
Question what is the `S` modifier doing in `S omnisharp` I assume just sorting right? I only ask because if I do those steps on a clean emacs I still get the same error. Here are my exact steps: 
1. Moved my .emacs.d file. 
2. Opened emacs and in scratch evaluated this code: 
```
(package-initialize)
(require 'package)
(add-to-list 'package-archives
             '("melpa" . "https://melpa.org/packages/"))
```
3. M-x package-refresh-contents
4. M-x package-list-packages
5. omnisharp/i/x

Afterwards I still get `omnisharp-server-management.el:64:1:Error: Wrong type argument: listp, value`. On the clean install for me it does come with the package `dash` with a `-doto` macro. If I delete just omnisharp and reinstall using the same steps I still get an error.

**Comment by razzmatazz** [comment]
oops, that was supposed to be `C-s` -- search

**Comment by razzmatazz** [comment]
@phoenixanimations what platform/emacs version are you on? so I could replace this 1:1?

**Comment by phoenixanimations** [comment]
I'm on OSX: 10.11.6, Emacs: 25.3. I did install and built it with Cocoa (not running it through a terminal), I also installed it using [emacs-plus](https://github.com/d12frosted/homebrew-emacs-plus). Not sure if that's necessary, but included it just in case.

**Comment by razzmatazz** [comment]
I believe I have fixed the installation problem in #403 and cannot replicate the problem anymore with the latest version.

The problem was byte compiliation error in omnisharp-server-management.el where the `-doto` macro and a `(lambda)` didn't interoperate somehow when byte-compiling (but worked when `(load`-ed ).

Please inform if it still fails for anyone.

**Comment by razzmatazz** [comment]
please reopen if this issue is still valid for anyone
